rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+1. In that same branch, add test cases for the new functionality to the module's testing configuration in `test/`
+1. Switch to the default branch (`master`/`main`), `cd` to the `test/` directory and run `terraform apply` to stand up the test resources
+1. Switch to the feature branch and run `terraform plan`/`terraform apply` to see the impact of the changes on the pre-existing resources
+1. When finished testing, run `terraform destroy` to tear down all test resources
+
+This workflow is intended to be carried out manually at first, but we might be able to use GitHub Actions to automate it at some point.
+
+#### Caveats for testing ECS applications
+
+In order to be able to test ECS applications, we will need to figure out how to allow the ECS service to retrieve a container image for deployment. There are a few blockers to this that will need to be overcome:
+
+- We'll need to override the ECR repo URL with the production AWS account's repo, and give the test account permission to pull images
+- We'll need to address an issue where apps can't start until the first GitHub Actions deploy because the template task definition references `latest`:
+  - Option 1: Add `latest` tag to all ECR images
+  - Option 2: Allow overriding image tag in module with [`.tfvars` files][terraform-tfvars].
+
+### Isolated Testing Environment
+
+As noted in the [guide-level explanation][guide-module-testing], testing will happen in the separate `mbta-ctd-test` AWS account. This account will be accessed using "assume role" permissions that full access to manage required AWS resources. The assumed role won't have full administrator access, but we'll manage the list of permitted services via IAM policies.
+
+## 3. Standardizing on Infrastructure Component Modules
+[reference-standardization]: #3-standardizing-on-infrastructure-component-modules
+
+As noted in the [guide-level explanation][guide-standardization], this is currently the most mature of the proposed improvements, in that the `ctd-ecs-app` and `ctd-rds-db` modules already exist and are seeing use. Aside from increasing adoption and carrying out the [testing workflow][guide-module-testing] and [module versioning][guide-module-versioning] recommendations, other improvements that will be beneficial include:
+
+- Handling all use cases for each module, using vars to enable/disable features that aren't likely to be used everywhere
+- Building in support for maintenance modes to minimize temporary configuration changes
+- Standardizing on the use of the `context` variable
+
+## 4. Module Versioning/Publishing
+[reference-module-versioning]: #4-module-versioningpublishing
+
+As noted in the [guide-level explanation][guide-module-versioning], module versioning and publishing is something we will set up for all infrastructure component modules, but which should be considered optional for application modules. See also the [corresponding subsection under rationale and alternatives][alternatives-module-versioning].
```

Comment:
we have a pretty big need for them on lamp.

 a good portion of the work we do is infrastructure R&D as we try to find the right combinations of infra and code solutions. we could probably achieve some of this with feature flags, but we could end up with a bunch of them, especially if we just want to run a test. versioning and only deploying versioned modules to production is our preferred setup.
 
i'm pretty sure (or as sure as i can be given i don't know most of the details) that versioning the ctd shared modules (like for RDS and ECS) is an important feature to work out so that teams can update their versions on their own schedule if breaking changes need to be applied.